helix-core/src/modeline.rs

```diff
@@ -0,0 +1,292 @@
+    language: Option<String>,
+    indent_style: Option<IndentStyle>,
+    line_ending: Option<LineEnding>,
+}
+
+impl Modeline {
+    pub fn parse(text: RopeSlice) -> Self {
+        let mut modeline = Self::default();
+
+        for line in text.lines().take(LINES_TO_CHECK).chain(
+            text.lines_at(text.len_lines())
+                .reversed()
+                .take(LINES_TO_CHECK),
+        ) {
+            // can't guarantee no extra copies, since we need to regex and
+            // regexes can't operate over chunks yet, but we can at least
+            // limit how much we potentially need to copy because modelines
+            // are typically quite short.
+            if line.len_chars() > LENGTH_TO_CHECK {
+                continue;
+            }
+            let line = Cow::<str>::from(line);
+            modeline.parse_from_line(&line);
+        }
+
+        modeline
+    }
+
+    pub fn language(&self) -> Option<&str> {
+        self.language.as_deref()
+    }
+
+    pub fn indent_style(&self) -> Option<IndentStyle> {
+        self.indent_style
+    }
+
+    pub fn line_ending(&self) -> Option<LineEnding> {
+        self.line_ending
+    }
+
+    fn parse_from_line(&mut self, line: &str) {
+        let mut saw_backslash = false;
+        let split_modeline = move |c| {
+            saw_backslash = match c {
+                ':' if !saw_backslash => return true,
+                '\\' => true,
+                _ => false,
+            };
+            c == ' ' || c == '\t'
+        };
+
+        if let Some(pos) = VIM_MODELINE_REGEX.find(line) {
+            for option in line[pos.end()..].split(split_modeline) {
+                let parts: Vec<_> = option.split('=').collect();
+                match parts[0] {
+                    "ft" | "filetype" => {
+                        if let Some(val) = parts.get(1) {
+                            self.language = Some(val.to_string());
+                        }
+                    }
+                    "sw" | "shiftwidth" => {
+                        if let Some(val) = parts.get(1).and_then(|val| val.parse().ok()) {
+                            if self.indent_style != Some(IndentStyle::Tabs) {
+                                self.indent_style = Some(IndentStyle::Spaces(val));
+                            }
+                        }
+                    }
+                    "ff" | "fileformat" => {
+                        if let Some(val) = parts.get(1) {
+                            self.line_ending = vim_ff_to_helix_line_ending(val);
+                        }
+                    }
+                    "noet" | "noexpandtab" => {
+                        self.indent_style = Some(IndentStyle::Tabs);
+                    }
+                    _ => {}
+                }
+            }
+        }
+
+        if let Some(pos) = HELIX_MODELINE_REGEX.find(line) {
+            let config = &line[pos.end()..];
+            match toml::from_str::<ModelineConfig>(config) {
```

Comment:
I dont think tonl is they syntax I would go for here. We are removing toml hopefzlly soon and generally it makes more sense to me what vim does and emulate typable commands (everything you allow users to set here has an interactive typable command that allows setting that config per document interactively). The parsing should be pretty simple. So for example it would be `// helix: line-ending=lf indent=4 lang=rust`

In terms of aliases you should also just support the aliases these commands support.